### DISCUSSION TITLE
Using multiple engines in the same project with sqlalchemy2

### DISCUSSION THREAD

**Discussion by JaviOverflow** [sqlalchemy/sqlalchemy#9975] [Usage Questions]
In older versions, I could bind an engine to a declarative_base instance like shown below so depending on which instance of those

```
from sqlalchemy import create_engine, text
from sqlalchemy.engine import URL
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker, Session

from environment import env

db_one__engine = create_engine(URL(drivername='mysql', **env['sqlalchemy']))
db_one__session_maker = sessionmaker(autocommit=False, autoflush=True, bind=db_one__engine)
db_one__base = declarative_base(bind=db_one__engine)

db_two__engine = create_engine(URL(drivername='mysql', **env['sce_soms__sqlalchemy']))
db_two__session_maker = sessionmaker(autocommit=False, autoflush=True, bind=db_two__engine)
db_two__base = declarative_base(bind=db_two__engine)
```

So if I created a new entity using db_one__base as a parent, it would go to that engine, while db_two__base would be mapped to the other engine/db.

How can I do this with sqlalchemy v2? I couldn't find the way to specify which engine each base should use.

**Comment by gordthompson** [comment]
Version 2 removed a number of bind= options that had evolved over time and eventually had become convoluted and potentially confusing. One of those options was to bind a declarative_base to an engine (which is effectively binding the engine to that object's MetaData(), and MetaData(bind=…) is also no longer permitted).

Now, we bind the engine to the *Session* in which we use the objects, not to the objects themselves:

```python
from sqlalchemy import create_engine, Column, Integer, String
from sqlalchemy.orm import declarative_base, Session

local_engine = create_engine("sqlite://")
remote_engine = create_engine("mssql+pyodbc://scott:tiger^5HHH@mssql_199")
Base = declarative_base()


class Thing(Base):
    __tablename__ = "thing"
    id = Column(Integer, primary_key=True)
    txt = Column(String(50))


Base.metadata.drop_all(local_engine)
Base.metadata.create_all(local_engine)
Base.metadata.drop_all(remote_engine)
Base.metadata.create_all(remote_engine)

with Session(local_engine) as local_session, local_session.begin():
    local_session.add(Thing(txt="for local database"))

with Session(remote_engine) as remote_session, remote_session.begin():
    remote_session.add(Thing(txt="for remote database"))

# check results

with local_engine.connect() as local_conn:
    print(local_conn.exec_driver_sql("SELECT * FROM thing").all())
    # [(1, 'for local database')]

with remote_engine.connect() as remote_conn:
    print(remote_conn.exec_driver_sql("SELECT * FROM thing").all())
    # [(1, 'for remote database')]
```

**Comment by zzzeek** [comment]
your example has two separate `sessionmaker` instances so you wouldnt be using the most common case for "bound metadata" which was so that a single `Session` could select from multiple engines, but if you wanted that you'd use the `binds` argument described at https://docs.sqlalchemy.org/en/20/orm/persistence_techniques.html#partitioning-strategies-e-g-multiple-database-backends-per-session .
